### PR TITLE
feat(nextcloud_test): Add local test target

### PR DESCRIPTION
**Pull request by provokateurin**
Closes https://github.com/nextcloud/neon/issues/1901

I can split this PR into multiple ones if desired.

Removing the maximum version check is done because major versions of apps don't necessarily indicate incompatible APIs. In fact, none of the existing APIs should ever have breaking changes except for cases when they are removed which also shouldn't happen.

I validated that the latest version of the server an all relevant apps work just fine with our current tests :tada:

### CODE REVIEW

File: packages/nextcloud_test/lib/src/test_target/test_target.dart [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+/// Creates a new [TestTargetFactory].
+TestTargetFactory createTestTargetFactory() {
+  return DockerContainerFactory();
+}
+
+/// Creates a new [NextcloudClient] for a given [username].
+///
+/// It is expected that the password of the user matches the its [username].
+FutureOr<NextcloudClient> createTestTargetClient(
```
Comment (Leptopoda) [inline packages/nextcloud_test/lib/src/test_target/test_target.dart:52]: Why isn't this a function of the `TestTargetInstance`?
I'd rather make it a super class that is extended, so the instance doesn't need to be passed into the call every thime.

File: packages/nextcloud_test/lib/src/test_target/test_target.dart [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  /// URL where the target is available from itself.
+  Uri get targetURL;
+
+  /// Creates an app password for [username] on the instance.
+  FutureOr<String> createAppPassword(String username);
+}
+
+/// Creates a new [TestTargetFactory].
+TestTargetFactory createTestTargetFactory() {
```
Comment (Leptopoda) [inline packages/nextcloud_test/lib/src/test_target/test_target.dart:39]: What is the benefit of this function?
I don't see how this helps.


Why not have this as a factory constructor in the `TestTargetFactory` class.

File: packages/nextcloud/doc/development.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+All tests should be standalone and never affect any other tests.
+With enough caution this is feasible to implement, but to verify the tests do not depend on the order of execution they should be run multiple times with different seeds:
+
+```sh
+for i in {0..4}; do fvm flutter test --test-randomize-ordering-seed $i; done
+```
+
+## Testing against development versions
+
+Make sure to have the following apps in your development setup and the appropriate versions you want to test checked out:
```
Comment (Leptopoda) [inline packages/nextcloud/doc/development.md:17]: > in your development setup

What do you mean by that?
That you need the git submodules checked out?

**Review by Leptopoda — COMMENTED** [review]
I really like the abstraction between the docker and local version :)
Nice work :+1:

File: packages/nextcloud/doc/development.md (hunk 3 above)
Comment (provokateurin) [reply packages/nextcloud/doc/development.md:17]: The setup when you develop on the server or an app for server. I'll clarify this

File: packages/nextcloud_test/lib/src/test_target/local.dart [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  void destroy() {}
+
+  @override
+  Uri hostURL = Uri.parse(Platform.environment['URL']!);
+
+  @override
+  late Uri targetURL = hostURL;
+
+  @override
+  FutureOr<String> createAppPassword(String username) async {
```
Comment (Leptopoda) [inline packages/nextcloud_test/lib/src/test_target/local.dart:89]: This always returns a future
```suggestion
  Future<String> createAppPassword(String username) async {
```

File: packages/nextcloud_test/lib/src/test_target/test_target.dart [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  /// URL where the target is available from itself.
+  Uri get targetURL;
+
+  /// Creates an app password for [username] on the instance.
+  FutureOr<String> createAppPassword(String username);
+
+  /// Creates a new [NextcloudClient] for a given [username].
+  ///
+  /// It is expected that the password of the user matches the its [username].
+  FutureOr<NextcloudClient> createClient({
```
Comment (Leptopoda) [inline packages/nextcloud_test/lib/src/test_target/test_target.dart:51]: This always returns a future
```suggestion
  Future<NextcloudClient> createClient({
```

**Review by Leptopoda — APPROVED** [review]

**Pull request merged**